hpke/aead.go

```diff
@@ -0,0 +1,80 @@
+package hpke
+
+import (
+	"crypto/cipher"
+	"errors"
+	"fmt"
+)
+
+type encdecCtx struct {
+	Suite
+	cipher.AEAD
+	baseNonce      []byte
+	seq            []byte
+	exporterSecret []byte
+}
+
+type sealCtx struct{ *encdecCtx }
+type openCtx struct{ *encdecCtx }
+
+func (c *encdecCtx) Export(expCtx []byte, length uint) []byte {
+	maxLength := uint(255 * c.KdfID.Hash().Size())
+	if length > maxLength {
+		panic(fmt.Errorf("size greater than %v", maxLength))
+	}
+	return c.labeledExpand(c.exporterSecret, []byte("sec"), expCtx, uint16(length))
+}
+
+func (c *encdecCtx) calcNonce() []byte {
+	nonce := (&[12]byte{})[:]
```

Comment:
could you please elaborate a bit.

Reply:
Sure. This assumes that the nonce length for this and all future AEADs is 12 bytes. That won't always be the case. I'm suggesting that the nonce length is a function of the AEAD in use. (Does that clarify?)

Reply:
```
	nonce := (&[12]byte{})[:]
	if len := c.NonceSize(); len != 12 {
		nonce = make([]byte, len)
	}
```
We fixed to 12 because it works for current AEADs and doesn't requires dinamic allocation. Also, the next line (if-then) prevents against the future AEADs with different nonce sizes.

Reply:
Sorry, I don't think I'm being clear. I'm saying that `12` should not be a magic number. It *depends on the AEAD in use*. (You can also statically allocate up to 16 bytes on the stack and then use however many are needed based on the AEAD, if you're worried about mallocs minimization.)

Reply:
I think I got the idea, take a look on the changes.